### ISSUE TITLE
Task sample won't work correctly

### ISSUE THREAD

**Issue by softlion** [dotnet/docs#9187]
The sample won't work correctly.

A task is not a thread. Instead, a task is scheduled on a thread by a taskscheduler.

In this sample, 100 tasks are created and run. The default taskscheduler creates 1 thread per CPU, no more. It then schedule each task on a free thread until there is no more task to run.

A task can be broken into 2 "runs" by the "await" keyword. I mean, when the await keyword is encountered, the taskscheduler frees the tread and schedule another task on it.
As the same thread is used for 2 running tasks, the lock keyword won't lock for calls originating from these threads.
Thus defeating its purpose.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b544eeb3-082e-1c67-eb76-f5ce0acee45e
* Version Independent ID: d4765a2e-ac77-e947-165c-5a0f88f4e8e2
* Content: [lock statement (C# Reference)](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/lock-statement?f1url=https%3A%2F%2Fmsdn.microsoft.com%2Fquery%2Fdev15.query%3FappId%3DDev15IDEF1%26l%3DEN-US%26k%3Dk(lock_CSharpKeyword)%3Bk(DevLang-csharp)%26rd%3Dtrue)
* Content Source: [docs/csharp/language-reference/keywords/lock-statement.md](https://github.com/dotnet/docs/blob/master/docs/csharp/language-reference/keywords/lock-statement.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

**Comment by pkulikov** [comment]
@softlion 
>As the same thread is used for 2 running tasks, the lock keyword won't lock for calls originating from these threads.

Could you please elaborate on that?

Also, what makes the sample in the article not work? Your reasoning in the second half of the post is based on `await` breaking the task in two parts. However, the article example doesn't use `await` at all.

Note that the docs say that:
>You can't use the `await` keyword in the body of a `lock` statement.

**Comment by softlion** [comment]
Yes, the problem arises when the method RandomlyUpdate is async, and you await inside it (try await Task.Delay(100) for example.

Of course as the method in the example is not async, it works. But it is VERY comfusing as all developers will try to use lock like in the credit/debit method, but may call them from an async method which has switch its context.
This means the lock won't work in all cases. Which leads to random errors in production.

Instead you should point to SemaphoreSlim, explaining that it resolves the problem.

**Comment by BillWagner** [comment]
I'm going to close this due to lack of response.

**Comment by softlion** [comment]
Sorry for the delay.
It won't work because tasks are NOT threads.

Tasks are queued and dispatched on max X threads where X is the number of processor minus one.

The lock keyword prevents only a code to run on different threads. Not on the same one. 

You forgot to use ConfigureAwait(false) in your await. This is required to simulate a task which returns to the queue while waiting on a resource.

So one task starts, is interrupted and goes back to the queue. Another one starts on the SAME THREAD and completes. Then the interrupted task continues.
The lock keyword did not stop the execution of the second task thus leading to bad behavior.

**Comment by svick** [comment]
@softlion That would only happen if you could `await` *inside* a `lock`, but [doing that is not allowed](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEAMACFBGArAbgFgAoElAJgwGESBvEjRjAewCMArCAY2AwBsMAXgwA7CAHcWHbsAAUASiLEmGBkxQAOLADYMAWQVrG9ZSqZ9mXANYZZfeUbMmzLrAE4dAOgAiEPgEMEWWxFRxUAX0dI4nCgA==) exactly for this reason.

If you use a `lock` before or after an `await`, then [there is no problem](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEAMACFBGArAbgFgAoElAJgwGESBvEjRjAewCMArCAY2AwBsMAXgwA7CAHcWHbsAAUASiLEmGBkxQAOLADYMAWQVrG9ZSqZ9mXANYZZfeUbMmzTAL6OVKAJw6AdABEIPgBDBFlsRUd3YlcgA==).

Also:

> Tasks are queued and dispatched on max X threads where X is the number of processor minus one.

That is not true, the ThreadPool will (slowly) create many more threads than that when necessary. For example, on my quad-core, `ThreadPool.GetMaxThreads` says it will create up to 1023 threads.
